teamcity: failed tests on master: acceptance/TestVersionUpgrade, acceptance/TestVersionUpgrade: TestVersionUpgrade/runMode=local, acceptance/TestVersionUpgrade, acceptance/TestVersionUpgrade: TestVersionUpgrade/runMode=local

The following tests appear to have failed:

[#525725](https://teamcity.cockroachdb.com/viewLog.html?buildId=525725):


```
--- FAIL: acceptance/TestVersionUpgrade (44.720s)
test_log_scope.go:81: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestVersionUpgrade437152343
test_log_scope.go:62: use -show-logs to present logs inline
--- FAIL: acceptance/TestVersionUpgrade: TestVersionUpgrade/runMode=local (44.720s)
version_upgrade_test.go:179: binary=v1.1.0
version_upgrade_test.go:179: cluster=1.0
version_upgrade_test.go:179: cluster=1.0-3
version_upgrade_test.go:179: binary=v1.1.1
version_upgrade_test.go:179: cluster=1.1
version_upgrade_test.go:179: binary=source
version_upgrade_test.go:179: cluster=1.1-6
version_upgrade_test.go:126: pq: result is ambiguous (error=initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure [propagate])
------- Stdout: -------
CockroachDB node starting at 2018-02-17 16:22:06.431671298 +0000 UTC
build:      CCL v1.0.6 @ 2017/09/14 15:15:48 (go1.8.3)
admin:      http://127.0.0.1:40937
sql:        postgresql://root@127.0.0.1:44285?sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/1
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/1
status:     initialized new cluster
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     1
CockroachDB node starting at 2018-02-17 16:22:06.74287676 +0000 UTC
build:      CCL v1.0.6 @ 2017/09/14 15:15:48 (go1.8.3)
admin:      http://127.0.0.1:44351
sql:        postgresql://root@127.0.0.1:37941?sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/2
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/2
status:     initialized new node, joined pre-existing cluster
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     2
CockroachDB node starting at 2018-02-17 16:22:06.766423447 +0000 UTC
build:      CCL v1.0.6 @ 2017/09/14 15:15:48 (go1.8.3)
admin:      http://127.0.0.1:43113
sql:        postgresql://root@127.0.0.1:39871?sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/3
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/3
status:     initialized new node, joined pre-existing cluster
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     3
CockroachDB node starting at 2018-02-17 16:22:12.967963785 +0000 UTC (took 4.6s)
build:      CCL v1.1.0 @ 2017/10/12 14:50:18 (go1.8.3)
admin:      http://127.0.0.1:34803
sql:        postgresql://root@127.0.0.1:40785?application_name=cockroach&sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/1
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/1
status:     restarted pre-existing node
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     1
CockroachDB node starting at 2018-02-17 16:22:15.100649543 +0000 UTC (took 0.7s)
build:      CCL v1.1.0 @ 2017/10/12 14:50:18 (go1.8.3)
admin:      http://127.0.0.1:37641
sql:        postgresql://root@127.0.0.1:43103?application_name=cockroach&sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/2
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/2
status:     restarted pre-existing node
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     2
CockroachDB node starting at 2018-02-17 16:22:17.14367291 +0000 UTC (took 0.7s)
build:      CCL v1.1.0 @ 2017/10/12 14:50:18 (go1.8.3)
admin:      http://127.0.0.1:39989
sql:        postgresql://root@127.0.0.1:45037?application_name=cockroach&sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/3
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/3
status:     restarted pre-existing node
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     3
CockroachDB node starting at 2018-02-17 16:22:26.460441239 +0000 UTC (took 5.3s)
build:      CCL v1.1.1 @ 2017/10/19 15:31:46 (go1.8.3)
admin:      http://127.0.0.1:42227
sql:        postgresql://root@127.0.0.1:37935?application_name=cockroach&sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/1
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/1
status:     restarted pre-existing node
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     1
CockroachDB node starting at 2018-02-17 16:22:28.390397307 +0000 UTC (took 0.7s)
build:      CCL v1.1.1 @ 2017/10/19 15:31:46 (go1.8.3)
admin:      http://127.0.0.1:43085
sql:        postgresql://root@127.0.0.1:42441?application_name=cockroach&sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/2
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/2
status:     restarted pre-existing node
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     2
CockroachDB node starting at 2018-02-17 16:22:30.333876718 +0000 UTC (took 0.6s)
build:      CCL v1.1.1 @ 2017/10/19 15:31:46 (go1.8.3)
admin:      http://127.0.0.1:41073
sql:        postgresql://root@127.0.0.1:38505?application_name=cockroach&sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/3
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/3
status:     restarted pre-existing node
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     3
CockroachDB node starting at 2018-02-17 16:22:44.822431265 +0000 UTC (took 6.9s)
build:               CCL v2.0-alpha.20180212-634-g2044e77 @ 2018/02/17 16:12:17 (go1.9.4)
admin:               http://127.0.0.1:36913
sql:                 postgresql://root@127.0.0.1:46069?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/1
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/1/cockroach-temp366559501
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/1/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/1
status:              restarted pre-existing node
clusterID:           b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:              1
CockroachDB node starting at 2018-02-17 16:22:46.690909137 +0000 UTC (took 0.7s)
build:               CCL v2.0-alpha.20180212-634-g2044e77 @ 2018/02/17 16:12:17 (go1.9.4)
admin:               http://127.0.0.1:36075
sql:                 postgresql://root@127.0.0.1:38271?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/2
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/2/cockroach-temp046374647
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/2/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/2
status:              restarted pre-existing node
clusterID:           b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:              2
CockroachDB node starting at 2018-02-17 16:22:48.606434818 +0000 UTC (took 0.7s)
build:               CCL v2.0-alpha.20180212-634-g2044e77 @ 2018/02/17 16:12:17 (go1.9.4)
admin:               http://127.0.0.1:36657
sql:                 postgresql://root@127.0.0.1:41565?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/3
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/3/cockroach-temp315620690
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/3/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/3
status:              restarted pre-existing node
clusterID:           b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:              3
test logs left over in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestVersionUpgrade437152343
--- FAIL: acceptance/TestVersionUpgrade (44.720s)
test_log_scope.go:81: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestVersionUpgrade437152343
test_log_scope.go:62: use -show-logs to present logs inline
--- FAIL: acceptance/TestVersionUpgrade: TestVersionUpgrade/runMode=local (44.720s)
version_upgrade_test.go:179: binary=v1.1.0
version_upgrade_test.go:179: cluster=1.0
version_upgrade_test.go:179: cluster=1.0-3
version_upgrade_test.go:179: binary=v1.1.1
version_upgrade_test.go:179: cluster=1.1
version_upgrade_test.go:179: binary=source
version_upgrade_test.go:179: cluster=1.1-6
version_upgrade_test.go:126: pq: result is ambiguous (error=initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure [propagate])
------- Stdout: -------
CockroachDB node starting at 2018-02-17 16:22:06.431671298 +0000 UTC
build:      CCL v1.0.6 @ 2017/09/14 15:15:48 (go1.8.3)
admin:      http://127.0.0.1:40937
sql:        postgresql://root@127.0.0.1:44285?sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/1
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/1
status:     initialized new cluster
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     1
CockroachDB node starting at 2018-02-17 16:22:06.74287676 +0000 UTC
build:      CCL v1.0.6 @ 2017/09/14 15:15:48 (go1.8.3)
admin:      http://127.0.0.1:44351
sql:        postgresql://root@127.0.0.1:37941?sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/2
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/2
status:     initialized new node, joined pre-existing cluster
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     2
CockroachDB node starting at 2018-02-17 16:22:06.766423447 +0000 UTC
build:      CCL v1.0.6 @ 2017/09/14 15:15:48 (go1.8.3)
admin:      http://127.0.0.1:43113
sql:        postgresql://root@127.0.0.1:39871?sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/3
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/3
status:     initialized new node, joined pre-existing cluster
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     3
CockroachDB node starting at 2018-02-17 16:22:12.967963785 +0000 UTC (took 4.6s)
build:      CCL v1.1.0 @ 2017/10/12 14:50:18 (go1.8.3)
admin:      http://127.0.0.1:34803
sql:        postgresql://root@127.0.0.1:40785?application_name=cockroach&sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/1
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/1
status:     restarted pre-existing node
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     1
CockroachDB node starting at 2018-02-17 16:22:15.100649543 +0000 UTC (took 0.7s)
build:      CCL v1.1.0 @ 2017/10/12 14:50:18 (go1.8.3)
admin:      http://127.0.0.1:37641
sql:        postgresql://root@127.0.0.1:43103?application_name=cockroach&sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/2
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/2
status:     restarted pre-existing node
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     2
CockroachDB node starting at 2018-02-17 16:22:17.14367291 +0000 UTC (took 0.7s)
build:      CCL v1.1.0 @ 2017/10/12 14:50:18 (go1.8.3)
admin:      http://127.0.0.1:39989
sql:        postgresql://root@127.0.0.1:45037?application_name=cockroach&sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/3
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/3
status:     restarted pre-existing node
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     3
CockroachDB node starting at 2018-02-17 16:22:26.460441239 +0000 UTC (took 5.3s)
build:      CCL v1.1.1 @ 2017/10/19 15:31:46 (go1.8.3)
admin:      http://127.0.0.1:42227
sql:        postgresql://root@127.0.0.1:37935?application_name=cockroach&sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/1
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/1
status:     restarted pre-existing node
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     1
CockroachDB node starting at 2018-02-17 16:22:28.390397307 +0000 UTC (took 0.7s)
build:      CCL v1.1.1 @ 2017/10/19 15:31:46 (go1.8.3)
admin:      http://127.0.0.1:43085
sql:        postgresql://root@127.0.0.1:42441?application_name=cockroach&sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/2
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/2
status:     restarted pre-existing node
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     2
CockroachDB node starting at 2018-02-17 16:22:30.333876718 +0000 UTC (took 0.6s)
build:      CCL v1.1.1 @ 2017/10/19 15:31:46 (go1.8.3)
admin:      http://127.0.0.1:41073
sql:        postgresql://root@127.0.0.1:38505?application_name=cockroach&sslmode=disable
logs:       /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/3
store[0]:   path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/3
status:     restarted pre-existing node
clusterID:  b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:     3
CockroachDB node starting at 2018-02-17 16:22:44.822431265 +0000 UTC (took 6.9s)
build:               CCL v2.0-alpha.20180212-634-g2044e77 @ 2018/02/17 16:12:17 (go1.9.4)
admin:               http://127.0.0.1:36913
sql:                 postgresql://root@127.0.0.1:46069?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/1
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/1/cockroach-temp366559501
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/1/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/1
status:              restarted pre-existing node
clusterID:           b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:              1
CockroachDB node starting at 2018-02-17 16:22:46.690909137 +0000 UTC (took 0.7s)
build:               CCL v2.0-alpha.20180212-634-g2044e77 @ 2018/02/17 16:12:17 (go1.9.4)
admin:               http://127.0.0.1:36075
sql:                 postgresql://root@127.0.0.1:38271?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/2
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/2/cockroach-temp046374647
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/2/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/2
status:              restarted pre-existing node
clusterID:           b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:              2
CockroachDB node starting at 2018-02-17 16:22:48.606434818 +0000 UTC (took 0.7s)
build:               CCL v2.0-alpha.20180212-634-g2044e77 @ 2018/02/17 16:12:17 (go1.9.4)
admin:               http://127.0.0.1:36657
sql:                 postgresql://root@127.0.0.1:41565?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestVersionUpgrade/runMode=local/3
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/3/cockroach-temp315620690
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/3/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster161849802/3
status:              restarted pre-existing node
clusterID:           b56ff315-7f10-430d-8e08-6a439327cb75
nodeID:              3
test logs left over in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestVersionUpgrade437152343

```

Please assign, take a look and update the issue accordingly.

--- a/pkg/acceptance/version_upgrade_test.go
+++ b/pkg/acceptance/version_upgrade_test.go
@@ -21,6 +21,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/pkg/errors"
 
 	"github.com/cockroachdb/cockroach/pkg/acceptance/cluster"
@@ -117,15 +118,35 @@ func (cv clusterVersionUpgrade) name() string { return fmt.Sprintf("cluster=%s",
 func (cv clusterVersionUpgrade) run(ctx context.Context, t *testing.T, c cluster.Cluster) {
 	t.Helper()
 
+	// hasShowSettingBug is true when we're working around
+	// https://github.com/cockroachdb/cockroach/issues/22796.
+	//
+	// The problem there is that `SHOW CLUSTER SETTING version` does not take into
+	// account the gossiped value of that setting but reads it straight from the
+	// KV store. This means that even though a node may report a certain version,
+	// it may not actually have processed it yet, which leads to illegal upgrades
+	// in this test. When this flag is set to true, we query
+	// `crdb_internal.cluster_settings` instead, which *does* take everything from
+	// Gossip.
+	v, err := roachpb.ParseVersion(string(cv))
+	if err != nil {
+		t.Fatal(err)
+	}
+	hasShowSettingBug := v.Less(roachpb.Version{Major: 1, Minor: 1, Unstable: 1})
+
 	func() {
 		db := makePGClient(t, c.PGUrl(ctx, rand.Intn(c.NumNodes())))
 		defer db.Close()
 
-		log.Infof(ctx, "upgrading cluster version to %s", string(cv))
-		if _, err := db.Exec(fmt.Sprintf(`SET CLUSTER SETTING version = '%s'`, string(cv))); err != nil {
+		log.Infof(ctx, "upgrading cluster version to %s", cv)
+		if _, err := db.Exec(fmt.Sprintf(`SET CLUSTER SETTING version = '%s'`, cv)); err != nil {
 			t.Fatal(err)
 		}
 	}()
+
+	if hasShowSettingBug {
+		log.Infof(ctx, "using workaround for upgrade to %s", cv)
+	}
 
 	for i := 0; i < c.NumNodes(); i++ {
 		testutils.SucceedsSoon(t, func() error {
@@ -133,22 +154,33 @@ func (cv clusterVersionUpgrade) run(ctx context.Context, t *testing.T, c cluster
 			defer db.Close()
 
 			var version string
-			if err := db.QueryRow("SHOW CLUSTER SETTING version").Scan(&version); err != nil {
-				t.Fatalf("%d: %s", i, err)
+			if !hasShowSettingBug {
+				if err := db.QueryRow("SHOW CLUSTER SETTING version").Scan(&version); err != nil {
+					t.Fatalf("%d: %s", i, err)
+				}
+			} else {
+				// This uses the receiving node's Gossip and as such allows us to verify that all of the
+				// nodes have gotten wind of the version bump.
+				if err := db.QueryRow(
+					`SELECT current_value FROM crdb_internal.cluster_settings WHERE name = 'version'`,
+				).Scan(&version); err != nil {
+					t.Fatalf("%d: %s", i, err)
+				}
 			}
 			if version != string(cv) {
-				return errors.Errorf("%d: expected version %s, got %s", i, string(cv), version)
+				return errors.Errorf("%d: expected version %s, got %s", i, cv, version)
 			}
 			return nil
 		})
 	}
+
+	log.Infof(ctx, "cluster is now at version %s", cv)
 
 	// TODO(nvanbenschoten): add upgrade qualification step.
 	time.Sleep(1 * time.Second)
 }
 
 func testVersionUpgrade(ctx context.Context, t *testing.T, cfg cluster.TestConfig) {
-	t.Skip("#22796")
 	steps := []versionStep{
 		binaryVersionUpgrade("v1.0.6"),
 		// v1.1.0 is the first binary version that knows about cluster versions,

--- a/pkg/sql/show_cluster_setting.go
+++ b/pkg/sql/show_cluster_setting.go
@@ -15,18 +15,73 @@
 package sql
 
 import (
+	"bytes"
 	"context"
 	"fmt"
 	"strings"
+	"time"
 
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/pkg/errors"
 
 	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/types"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
 	"github.com/cockroachdb/cockroach/pkg/util/duration"
 )
+
+func (p *planner) showStateMachineSetting(
+	ctx context.Context, st *cluster.Settings, s *settings.StateMachineSetting, name string,
+) (*tree.DString, error) {
+	var d *tree.DString
+	// For statemachine settings (at the time of writing, this is only the cluster version setting)
+	// we show the value from the KV store and additionally wait for the local Gossip instance to
+	// have observed the value as well. This makes sure that cluster version bumps become visible
+	// immediately while at the same time guaranteeing that a node reporting a certain version has
+	// also processed the corresponding Gossip update (which is important as only then does the node
+	// update its persisted state; see #22796).
+	if err := retry.ForDuration(10*time.Second, func() error {
+		datums, err := p.QueryRow(ctx, "SELECT value FROM system.settings WHERE name = $1", name)
+		if err != nil {
+			return err
+		}
+		var prevRawVal []byte
+		if len(datums) != 0 {
+			dStr, ok := datums[0].(*tree.DString)
+			if !ok {
+				return errors.New("the existing value is not a string")
+			}
+			prevRawVal = []byte(string(*dStr))
+		}
+		// Note that if no entry is found, we pretend that an entry
+		// exists which is the version used for the running binary. This
+		// may not be 100.00% correct, but it will do. The input is
+		// checked more thoroughly when a user tries to change the
+		// value, and the corresponding sql migration that makes sure
+		// the above select finds something usually runs pretty quickly
+		// when the cluster is bootstrapped.
+		kvRawVal, obj, err := s.Validate(&st.SV, prevRawVal, nil)
+		if err != nil {
+			return errors.Errorf("unable to read existing value: %s", err)
+		}
+
+		// NB: if there is no persisted cluster version yet, this will match
+		// kvRawVal (which is taken from `st.SV` in this case too).
+		gossipRawVal := []byte(s.Get(&st.SV))
+		if !bytes.Equal(gossipRawVal, kvRawVal) {
+			return errors.Errorf("gossip and KV store disagree about value")
+		}
+
+		d = tree.NewDString(obj.(fmt.Stringer).String())
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return d, nil
+}
 
 func (p *planner) ShowClusterSetting(
 	ctx context.Context, n *tree.ShowClusterSetting,
@@ -76,32 +131,11 @@ func (p *planner) ShowClusterSetting(
 			case *settings.StringSetting:
 				d = tree.NewDString(s.String(&st.SV))
 			case *settings.StateMachineSetting:
-				// Show consistent values for statemachine settings. This isn't necessary
-				// for correctness, but helpful for testability.
-				datums, err := p.QueryRow(ctx, "SELECT value FROM system.settings WHERE name = $1", name)
+				var err error
+				d, err = p.showStateMachineSetting(ctx, st, s, name)
 				if err != nil {
 					return nil, err
 				}
-				var prevRawVal []byte
-				if len(datums) != 0 {
-					dStr, ok := datums[0].(*tree.DString)
-					if !ok {
-						return nil, errors.New("the existing value is not a string")
-					}
-					prevRawVal = []byte(string(*dStr))
-				}
-				// Note that if no entry is found, we pretend that an entry
-				// exists which is the version used for the running binary. This
-				// may not be 100.00% correct, but it will do. The input is
-				// checked more thoroughly when a user tries to change the
-				// value, and the corresponding sql migration that makes sure
-				// the above select finds something usually runs pretty quickly
-				// when the cluster is bootstrapped.
-				_, obj, err := s.Validate(&st.SV, prevRawVal, nil)
-				if err != nil {
-					return nil, errors.Errorf("unable to read existing value: %s", err)
-				}
-				d = tree.NewDString(obj.(fmt.Stringer).String())
 			case *settings.BoolSetting:
 				d = tree.MakeDBool(tree.DBool(s.Get(&st.SV)))
 			case *settings.FloatSetting:

